backup_prog.1
.\"                              hey, Emacs:   -*- nroff -*-
.\"
.\"
.TH BACKUP_PROG 1 "October 29, 2009"
.\" Please update the above date whenever this man page is modified.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins (default)
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
.B backup_prog \- a perl/curses based backup program.
.SH SYNOPSIS
.B backup_prog
.RI [ options ]
.br
.SH DESCRIPTION
\fBbackup_prog\fP is a perl/curses based program used to backup my 
development data. It provide many backup type (daily/weekly...) and
support distant retrieving of files using scp.
.PP
For more informations on \fBbackup_prog\fP you can see the info node : 
\fBinfo backup_prog\fP or the perldoc documentation : \fBperldoc 
backup_prog.pl\fP.

.SH OPTIONS
\fBbackup_prog\fP accepts the following command-line options.
.\" ===============================================
.\"          
.\" ===============================================
.SS GENERAL OPTIONS
Here are the options I can put in other sections. Mandatory arguments to long
options are mandatory for short options too.
.TP
.\" ================ --config-file option
NYI \fB-c\fP, \fB\-\-config\-file\fP \fIfilename\fP
Set the config file to a named one. Disable parsing of all others
system wide config files. Using this option can speed up processing
of configuration files.
.TP
.\" ================ --help option
NYI \fB-h\fP, \fB\-\-help\fP
Show a list of command line options and exit.
.TP
.\" ================ --backup-type option
NYI \fB-t\fP, \fB\-\-backup\-type\fP \fIhourly\fP|\fIdaily\fP|\fIweekly\fP|\fIcomplete\fP
.RS
.TP .4i
\fBhourly\fP :
This backup type intended to eb used with the \fB--no-gui\fB option, eventually
cron-ed. It was made to allow weekly backup even if some hosts are unreachable.
.TP
\fBdaily\fP : 
This is a backup type only provided for consummers who wants daily backup. 
Do not used for my personnal datas.
.TP
\fBweekly\fP : 
This is the most used backup type, intended to be placed on small cartridge,
most of the big files are removed (data files, PDFs...). It size is around
28 Mo.
.TP
\fBcomplete\fP : 
The biggest backup, containing files removed during the \fBweekly\fP one. This
backup intended to be placed on a CD and is about 300Mo actually.
.RE
.TP
.\" ================ --no-gui option
NYI \fB\-\-no-gui\fP
Do not draw the curses-based graphical user interface. This option is used
to launch \fBbackup_prog\fP as a background process.
.\" ===============================================
.\"          
.\" ===============================================
.SS LOGGER OPTIONS
.TP
.\" ================ --log-level option
NYI \fB\-\-log\-level\fP \fIinfo\fP|\fIwarn\fP|\fIerror\fP
Set the logging level to \fBlevel\fP, all logger messages with inferior
level will not be printed. \fBlevel\fP can be one of the following :
.RS
.TP .4i
\fBinfo\fP :
All messages, even informative one will be logged. The info level also contains
messages relative to configuration files and command line arguments parsing.
.TP
\fBwarn\fP :
Only warning and error messages are logged. Informative ones are omitted.
.TP
\fBerror\fP : 
Only error messages are logged. Warning and informative ones are omitted.
.RE
.TP
.\" ================ --log-file option
NYI \fB\-\-log\-file\fP \fIfilename\fP
Change the files where are printed logger messages to \fIfilename\fP. By 
default, the logfile is \fBbackup_prog.log\fP (in the execution directory).
.TP
.\" ================ --log-ident-len option
NYI \fB\-\-log\-ident\-len\fP \fI-1\fP|\fIn\fP
The Logger has the ability to truncate the Perl packages name in the log file.
The possible values are :
.RS
.TP .4i
\fB-1\fP :
Do not truncate at all the identification part of the logged message. Please
note that this option also disable the justify ability of the logger : the
II, WW and other indications will not be aligned anymore.
.TP
\fBn\fP :
Truncate on the left the identification part to \fBn\fP characters.
.RE
.TP
.\" ================ --log-mode option
NYI \fB\-\-log\-mode\fP \fIappend\fP|\fInewfile\fP
Defines how the logger will deal with the log file :
.RS
.TP .4i
\fBappend\fP :
The messages, including introduction, will be append to the logfile if it
exists.
.TP
\fBnewfile\fP :
The old logfile will be overwritten if it exists. 
.RE
.\" ==================================
.SH HISTORY
\fBbackup_prog\fP started when I saw how it was awful to maintain my old 
backup_prog.sh shell script (700+ SLOC).
.\" ==================================
.SH REPORTING BUG
Report bugs to  <rainbru@free.fr>.


ChangeLog
* v0.0.1-17 (09 Sep. 2022 - 21 Sep. 2022) SLOC 347 PO 37/49
- examples/README update
- First ActionList drawActions implementation
- Fix an unintended interpolation in ActionList implementation
- Implement len and append methods for ActionList
- Start draw_label overridind in ElepsedTime widget
- Add an example test example
- ActionList is a basic box
- Better ElapsedTime widget placement

* v0.0.1-16 (30 Aug. 2022 - 09 Sep. 2022) SLOC 312 PO 37/49
- Unit tests source cleanup
- Better information message for check-target-dir Makefile rule
- Add a m4 existence check in gen-makefiles.sh
- Fix a nbsp charcater in echo command in check-target_dir rule
- Add check-target-dir makefile rule to m4 generation code
- Close the UI-related ROADMAP section
- Widget::get_label implementation
- Add pod documentation for ElaspedTime widget

* v0.0.1-15 (15 Aug. 2022 - 30 Aug. 2022) SLOC 172 PO 37/49
- Implement the get_elapsed_str() method
- Start to print time in logger
- ElapsedTime call parent constructor
- Start ElapsedTime widget implementation
- Implement WidgetList::force_refresh()
- Add WidgetList refresh implementation and fix internal widget (was 0)
- Implement WidgetList as a package our variable
- Add WidgetList::append unit test

* v0.0.1-14 (29 Jul. 2022 - 15 Aug. 2022) SLOC 329 PO 37/49
- Add first WidgetList unit tests
- WidgetList empty class added
- Widget::draw_border is now an instance function
- Add version widget and use text_y for centered labels
- Widget::draw_label is an instance method
- Fix the Widget label issue and add a warning for bad label_y values
- Widget uses Exception in case of bad Alignment flag
- Add perl-libintl-perl dependency to fix gettext shorthand issue

* v0.0.1-13 (07 Jui. 2022 - 29 Jul. 2022) SLOC 325 PO 37/49
- Implement widget::set_label()
- Remove doc/NEEDED_MODULES (outdated and install explained in README)
- Explain 'polyglotman/rman' instalation in README
- Implement Widget's dirty flag
- Finally fixed the widget drawing issue
- Improve i18n documentation
- Create i18n share directory before installing .mo files
- Fix the undefined setlocale error

* v0.0.1-12 (06 may 2010 - 07 Jui. 2022) SLOC ??? PO ??/??
- 5 Jul 2022: get back after a 12y pause
- Widget.pm handles all label alignment (in draw_label function)
- Widget.pm text alignment constants definition
- Widget.pm handles options through a hashmap
- UserInterface/Widget.pm implementation started


backup_prog.pl
#!/usr/bin/perl -w

#use strict;

# Fix the setlocale() error
use POSIX qw(locale_h);
use locale;

use Curses;
use Locale::gettext qw(gettext textdomain); 
use Locale::TextDomain ('backup_prog');

#use POSIX; # Got a 'Subroutine main::getchar redefined' error
use Error qw(:try);

use BackupProg::UserInterface::MainWindow;
use BackupProg::Common::Logger qw(:LogLevels :LogModes);
use BackupProg::Parser::CommandLine;
use BackupProg::Common::Config;

# First, open the Logger
my $log;
try{
    $log = BackupProg::Common::Logger
	->instance( "backup_prog.log", LL_INFO, 25, LM_NEWFILE );
}
catch BackupProg::Exception::BadArgument with{
    my $E = shift;
    print gettext("Error creating Logger : ").$E->message()."\n";
}
catch BackupProg::Exception::OpenFile with{
    my $E = shift;
    print gettext("Error opening logfile : ").$E->message()."\n";
};

=for Config files test
use Data::Dump qw(dump);
use Config::General;
#my $conf = new Config::General("get_distant.d/test");
my $conf = new Config::General("doc/SKINS");
my %config = $conf->getall();
dump(%config);
=cut

# Trying GNU gettext
$log->LOGI("Setting up GNU gettext");
setlocale(LC_MESSAGES, "");
textdomain("BackupProg");

# TRANSLATOR : This is a message for You
print gettext("Welcome to backup_prog\n");

# Creating configuration object
my $conf = BackupProg::Common::Config->new();

# Command-line arguments parsing
my $clp = BackupProg::Parser::CommandLine->new();
$clp->parse_before_configfile();
$log->LOGI("Parsing config file");
$clp->parse();

#exit 0; # WARNING =======================================================

$log->LOGE("Should be written in the temp file");

$log->configure($conf);

use BackupProg::Parser::ConfigFile;
my $cf = BackupProg::Parser::ConfigFile->new();

try{
    my $mw = BackupProg::UserInterface::MainWindow->new();
}
catch BackupProg::Exception::BadArgument with{
    print "Catched a MainWindow exception\n";
    my $E = shift;
    $log->LOGX($E);
};
# Close Logger
$log->close(); 

1;
=pod

=head1 NAME

baclkup_prog - A perl/curses program used to backup files.

=head1 SYNOPSIS

backup_prog [options]

  Options:
    -h, --help			print a brief help message
    -c, --config-file <file>	set the configuration file to be used
=cut
